src/server/worker/datum/iterator.go

```diff
@@ -312,6 +317,72 @@ func (d *crossIterator) DatumN(n int) []*common.Input {
 	return result
 }
 
+type groupIterator struct {
+	datums   [][]*common.Input
+	location int
+}
+
+func newGroupIterator(pachClient *client.APIClient, group []*pps.Input) (Iterator, error) {
+	groupMap := make(map[string][]*common.Input)
+	keys := make([]string, 0, len(group))
+	result := &groupIterator{}
+	defer result.Reset()
+
+	// okay, so we have a slice of pps Inputs
+	for _, input := range group {
+		// turn our inputs into iterators
+		datumIterator, err := NewIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		// iterate through each iterator to get the individual datums
+		for datumIterator.Next() {
+			datum := datumIterator.Datum()
+			for _, datumInput := range datum {
+				// put the datums in an map keyed by GroupBy
+				groupDatum, ok := groupMap[datumInput.GroupBy]
+				if !ok || groupDatum == nil {
```

Comment:
Is it necessary to have both of these conditions? It seems like they should always both be true or always both be false in which case it's slightly confusing to be checking both.